cgi-bin/LJ/Web.pm

```diff
@@ -752,53 +752,64 @@ sub check_referer {
 }
 
 # <LJFUNC>
-# name: LJ::icons_for_remote
+# name: LJ::icon_keyword_menu
 # class: web
-# des: Gets all the userpics for a given user, including an item for the default userpic.
-# args: remote
-# des-remote: a user object.
-# returns: An array of {value => ..., text => ..., data => { url => ...}}
-#          hashrefs, suitable for use as the items in an LJ::html_select().
-#          If no userpics or if remote is undefined, an empty array.
+# des: Gets all userpics for a given user, separated by keyword. Use this when
+#      building a drop-down icons menu.
+# args: user
+# des-user: a user object.
+# returns: An array of hashrefs like:
+#          {value => ..., text => ..., data => { url => ..., description => ... }}
+#          which can be passed directly to an LJ::html_select(). Includes an
+#          item for each keyword (thus duplicating icons with multiple
+#          keywords), and an item for the default userpic. If no userpics or if
+#          user is undefined, returns an empty array.
 # </LJFUNC>
-sub icons_for_remote {
-    my ($remote) = @_;
-    my @pics;
+sub icon_keyword_menu {
+    my ($user) = @_;
 
-    {
-        my %res;
-        if ($remote) {
-            LJ::do_request(
-                {
-                    mode         => "login",
-                    ver          => $LJ::PROTOCOL_VER,
-                    user         => $remote->user,
-                    getpickws    => 1,
-                    getpickwurls => 1,
-                },
-                \%res,
-                { noauth => 1, userid => $remote->userid }
-            );
-        }
+    return () unless ($user);
 
-        if ( $res{pickw_count} ) {
-            for ( my $i = 1 ; $i <= $res{pickw_count} ; $i++ ) {
-                push @pics, [ $res{"pickw_$i"}, $res{"pickwurl_$i"} ];
-            }
-            @pics = sort { lc( $a->[0] ) cmp lc( $b->[0] ) } @pics;
-            my $defaultpicurl = $res{defaultpicurl}
-                || ( $LJ::IMGPREFIX . $LJ::Img::img{nouserpic_sitescheme}->{src} );
-            @pics = (
-                {
-                    value => "",
-                    text  => LJ::Lang::ml('/talkpost.bml.opt.defpic'),
-                    data  => { url => $defaultpicurl }
+    my @icons = grep { !( $_->inactive || $_->expunged ) } LJ::Userpic->load_user_userpics($user);
+
+    if (@icons) {
```

Comment:
nit: Slightly cleaner (less indenting) if this is,

```perl
my @icons = ...;
return () unless @icons;

@icons = LJ::Userpic->...
return @icons;
```

Reply:
Yeah, ok! 

If I'm doing that, maybe I should also just return that `(default-icon, map-expression)` list directly, instead of re-assigning `@icons` right before returning it.